### PR TITLE
feat: in-process Jedi semantic frontend for Python behind PYTHON_FRONTEND

### PR DESCRIPTION
**Pull request by vitali87**
Stage 1 of #1183 (the issue stays open for stage 2, the opt-in Pyright tier). The ironic gap closes: Python, which gets the deepest FLOWS_TO walk in the repo, gains its first semantic layer — in-process, zero toolchain.

## What's in

- **`parsers/py_frontend/`**: the Jedi fact provider behind the `LanguageFrontend` protocol (#1178). A stdlib-`ast` walker enumerates the call sites worth a semantic query — attribute calls and import-bound bare calls; module-local bare calls stay on the heuristics' home turf — with byte-exact name-token positions matching the tree-sitter `CallSiteKey` convention (ast offsets are UTF-8 bytes; jedi speaks char columns, so both directions convert per line). Each site gets `Script.infer()`: a single first-party inference becomes a `resolved_call_sites` fact targeted at the definition's Pass-2 span key (the def/class keyword column, i.e. the line's leading-whitespace width); a resolution outside the repo (stdlib, site-packages, builtins) becomes an `external_sites` proof. Ambiguity (zero or multiple inferences) emits nothing, never a guess.
- **Consumption**, mirroring the Go shape exactly: shared-reference fact maps on `DefinitionProcessor`, threaded through `TypeInferenceEngine`, consulted by `CallResolver.resolve_python_call_site` (join via the shared `call_site_key`/`declared_location` helpers), with a `PY_EXTERNAL_TARGET` sentinel in the call processor that suppresses trie fabrication; any miss falls back to today's Python heuristics with `call_point` preserved.
- **`PYTHON_FRONTEND` setting** (`heuristic` default | `jedi`): opt-in for now, and `jedi`-when-uninstalled degrades to heuristics with a warning. The parser fingerprint records the RESOLVED mode so jedi and non-jedi graphs never share an identity.
- **Cost control**: one `jedi.Project` shared across files; a 2s per-file budget that degrades that file to heuristics (logged) rather than stalling the index.
- **Realtime**: `.py` changes re-run the frontend on the watch path (the #1319 posture), since the facts are position-keyed against the repo-wide import graph.
- **Dependency**: `jedi>=0.19` under a new `python-semantics` extra; `uv.lock` regenerated minimally (jedi + parso).

## The discriminating case

`pkg/__init__.py: from .impl import real_fn as f` + `from pkg import f; f()`: the heuristics emit NO call edge at all; jedi follows the aliased re-export to the definition and the graph gains `caller.use -> pkg.impl.real_fn`. Test-pinned in both directions (`test_aliased_reexport_is_the_jedi_discriminator`, `test_heuristic_mode_stays_heuristic`).

## Honesty notes

- A repo-local `os.py` next to the caller resolves as the call target — that is what `import os` binds at runtime (script-dir precedence), so the edge is correct, and it's pinned as such (`test_repo_local_shadow_beats_the_stdlib`).
- Off (the default), the resolver short-circuits on empty fact maps: the resolution battery (923 tests across python/resolver/calls/realtime) is green unchanged.

## Tests

`test_python_frontend.py` (jedi-gated via importorskip): the discriminator pair; stdlib external proof at the fact level; decorated-function resolution; repo-local shadow semantics; budget-exhaustion degrade; ast-walker byte columns on unicode lines; byte/char column round-trip. All 9 green; battery 923 green.

## Summary by CodeRabbit

* **New Features**
  * Added optional semantic Python analysis using Jedi for more accurate call resolution.
  * Added configurable heuristic (default) and Jedi analysis modes.
  * Improved distinction between repository calls and external library calls.
  * Python file updates now refresh semantic call relationships automatically.
  * Added fallback safeguards when semantic analysis is unavailable or exceeds its time budget.

* **Bug Fixes**
  * Improved handling of re-exports, aliases, decorated functions, Unicode source, and local modules shadowing standard-library modules.

### CODE REVIEW

File: codebase_rag/parsers/py_frontend/frontend.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            continue
+        target = _target_of(names, repo_path)
+        if target is None:
+            continue
+        key: CallSiteKey = (rel_file, line, byte_col, simple_name)
+        module_path = target.module_path
+        if module_path is None or not str(module_path).startswith(str(repo_path) + "/"):
+            facts.external_sites.add(key)
+            continue
+        target_rel = Path(module_path).relative_to(repo_path).as_posix()
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/py_frontend/frontend.py]: Fixed in 09be0c10: containment uses Path.is_relative_to (and relative_to for the rel path), so Windows-native separators and edge roots classify correctly; the string prefix check is gone.

File: codebase_rag/parsers/py_frontend/frontend.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    import jedi
+
+    for simple_name, line, byte_col in sites:
+        if time.monotonic() > deadline:
+            return False
+        if line - 1 >= len(source_lines):
+            continue
+        char_col = _byte_to_char_col(source_lines[line - 1], byte_col)
+        try:
+            names = script.infer(line, char_col)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/py_frontend/frontend.py]: Fixed in 09be0c10: _resolve_file collects into a file-local SemanticFacts and merges only after a FINAL deadline check, so a slow last infer() discards the whole file's facts instead of recording them, and degradation is atomic (no half-resolved prefix). jedi has no cancellation API, so a single in-flight infer can still overshoot the wall clock by its own duration; the docstring states that bound explicitly. A cancellable process boundary is the Pyright tier's natural home (stage 2 runs out-of-process anyway) and is out of scope for the in-process stage.

File: codebase_rag/parsers/py_frontend/frontend.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+
+def _char_to_byte_col(line_text: str, char_col: int) -> int:
+    return len(line_text[:char_col].encode(cs.ENCODING_UTF8))
+
+
+def _target_of(names, repo_path: Path):
+    if len(names) != 1:
+        return None
+    name = names[0]
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/py_frontend/frontend.py:113]: Renamed in 09be0c10: _target_of is now _single_resolvable_target (with the ceiling contract stated in its comment) and the test helper is _calls_edges_for.

File: codebase_rag/parsers/py_frontend/frontend.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            continue
+        # The Pass-2 span index keys Python definitions at the def/class
+        # KEYWORD (the line's first code byte), not the name token jedi
+        # reports; the indent is ASCII so its byte and char widths agree.
+        def_line = target_lines[target.line - 1]
+        target_byte_col = len(def_line) - len(def_line.lstrip())
+        facts.resolved_call_sites[key] = ResolvedCallSite(
+            simple_name, target_rel, target.line, target_byte_col
+        )
+    return True
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/py_frontend/frontend.py]: Both parts fixed in 09be0c10: (1) file-local SemanticFacts merged only after a final deadline check — a budget-blowing file contributes nothing, and a slow final infer() is discarded rather than recorded (the docstring states the one-call overshoot bound; a cancellable process boundary belongs to the out-of-process Pyright tier in stage 2); (2) containment now uses Path.is_relative_to / relative_to, covering Windows separators and root-repo edge cases. The same commit also fixes the CI Type Check failure by carrying jedi in the test extra (every CI job already syncs it), which additionally makes the jedi-gated tests RUN in CI instead of skipping.
